apis/Google.Cloud.PubSub.V1/Google.Cloud.PubSub.V1/AckError.cs

```diff
@@ -0,0 +1,127 @@
+using System.Collections.Generic;
+using System.Linq;
+
+namespace Google.Cloud.PubSub.V1;
+
+/// <summary>
+/// This class contains the temporary and/or permanent failures from Acknowledge and ModifyAckDeadline RPCs for exactly once subscriptions.
+/// </summary>
+internal sealed class AckError
+{
+    internal AckError(IEnumerable<KeyValuePair<string, string>> temporaryFailures, IEnumerable<KeyValuePair<string, string>> permanentFailures)
+    {
+        TemporaryFailures = temporaryFailures;
+        PermanentFailures = permanentFailures;
+    }
+
+    /// <summary>
+    /// True if there are any permanent failures else false if <see cref="PermanentFailureIds"/> is empty.
+    /// </summary>
+    internal bool HasPermanentFailures => PermanentFailureIds.Any();
+
+    /// <summary>
+    /// Gets the AckIds/ExtendIds/NackIds with temporary failures.
+    /// This collection will be populated with AckIds/ExtendIds/NackIds only if the failure is temporary,
+    /// else this collection will be empty (but not null).
+    /// </summary>
+    internal IEnumerable<string> TemporaryFailureIds => TemporaryFailures.Select(kvp => kvp.Key);
+
+    /// <summary>
+    /// Gets the temporary failures.
+    /// This collection will be populated with AckIds/NackIds/ExtendIds with their corresponding failures only if the failure is temporary,
+    /// else this collection will be empty (but not null).
+    /// </summary>
+    internal IEnumerable<KeyValuePair<string, string>> TemporaryFailures { get; }
+
+    /// <summary>
+    /// Gets the AckIds/ExtendIds/NackIds with permanent failures.
+    /// This collection will be populated with AckIds/ExtendIds/NackIds only if the failure is permanent,
+    /// else this collection will be empty (but not null).
+    /// </summary>
+    internal IEnumerable<string> PermanentFailureIds => PermanentFailures.Select(kvp => kvp.Key);
+
+    /// <summary>
+    /// Gets the permanent failures.
+    /// This collection will be populated with AckIds/NackIds/ExtendIds with their corresponding failures only if the failure is permanent,
+    /// else this collection will be empty (but not null).
+    /// </summary>
+    internal IEnumerable<KeyValuePair<string, string>> PermanentFailures { get; }
+
+    /// <summary>
+    /// Attempts to get the exception from the current instance of <seealso cref="AckError"/>.
+    /// </summary>
+    /// <remarks>
+    /// This method will return <c>true</c> if and only if there are permanent failures
+    /// in the current instance of <seealso cref="AckError"/>,
+    /// in which case <paramref name="exception"/> will be populated.
+    /// </remarks>
+    /// <param name="exception">The <see cref="AggregateException"/> to populate if there are permanent failures.</param>
+    /// <returns><c>true</c> if there are permanent failures and <paramref name="exception"/> is populated, <c>false</c> otherwise.</returns>
+    internal bool TryGetAcknowledgementException(out AggregateException exception)
```

Comment:
Possibly TryGetPermanentException? And possibly just return the exception itself or null, instead of using `out`?
(That's what I personally tend to do when I'm returning something nullable.)